### DISCUSSION TITLE
Thanks so much and whatabout compression options?

### DISCUSSION THREAD

**Discussion by paulf123b** [UnconnectedBedna/shrink-backup#49] [Ideas]
Hi. Thank you so much for building this. I've messed around with all of these: plain old dd, rsync, raspiBackup, PiShrink and other scripts but I like your solution the most! I use it to keep my 3D printer's Pi backed up.

I know this would be complicated with the incremental backups, but personally I'd prefer to have full dailys anyway.... but yeah have you any thoughts of a compression option(s?) I know I can manually or by script compress but an integrated method would be great. With 7zip I compress down to about 1/3 of the .img. With my skills not much stronger than copy/paste I did tried piping the output ( | 7z a -si ) but of course this was silly given the scripts prompts and output.

Thanks again!

**Comment by UnconnectedBedna** [comment]
Hello and thank you so much for the very kind words!
Makes me really glad to read!

I have thought about compression and decided not to do it.
Reasoning:
Just as you state incremental backups would take WAY longer.
Fixing an uncompression of a file before the backup runs is not a problem per se, but the idea of the script is very much "KISS", keep it stupid simple, so when I started to write the code, it just felt wrong. "This is for the user to do if they want".
Beside that, it also introduces yet another dependency in form of compression software.
Arguably tar is _usually_ installed, but not always, so I would have to check for that too.

In the end I just felt that the amount of work, compared to how easy it is for the user to do the same, was not worth it to me.

If you want a workflow where this is achieved, I suggest writing your own script that launches shrink-backup and when done, compresses the image for you.
Or if you prefer something like crontab, just make sure to first run shrink-backup then add subsequent commands separated with `&&`.

But as always, if someone want to create this, and it makes sense to include, I have nothing against accepting a pr that compresses the image after it is created.
But I would also demand that the user could then run an incremental backup towards that compressed file. :slightly_smiling_face:

**Comment by UnconnectedBedna** [comment]
Here is a code snippet you can use to start the script, and then add a compression command before the unmounting of the storage location (if you want to use that)

Could for example be called `backup.sh`
Change the path to `shrink-backup` in each of the options in the `case "$1" in` to where you have the script located. (or remove the path if `shrink-backup` is in your `$PATH`)
The img file is hardcoded in this snippet (`backup.img`), but you can obv change that to your liking, maybe by using `date` or something similar.
```
#!/usr/bin/env bash

# This whole function can be removed/commented out if you do not mount the backup location using this script
# This function ensures the backup locations is unmounted if you press n (in script confirmation) or ctrl+c during the script
# Make sure path is correct to where your mount is located
# The space after the path is intentional, ie NOT: '/mnt/backup', but rather: '/mnt/backup ' is REQUIRED
function cleanup() {
  if grep -qs '/mnt/backup ' /proc/mounts; then
    echo '## Unmounting NFS backup'
    umount /mnt/backup
  fi
}
trap cleanup EXIT SIGTERM SIGINT

# Check if script is runing as root
if [ "$EUID" -ne 0 ]; then
  echo '!! THIS SCRIPT MUST BE RUNNING AS ROOT! (WITH SUDO)'
  exit 1
fi

# If you want the script to mount the backup location, NFS (192.168.1.100 at /mnt/backup) in this example.
# Change to the method you mount or comment out/remove if not wanted
echo '## Mounting NFS backup'
mount -t nfs 192.168.1.100:/foo/bar /mnt/backup
sleep 1

# Extra check if the backup location is mounted, comment out/remove if not wanted
# Make sure path is correct to where your mount is located
# The space after the path is intentional, ie NOT: '/mnt/backup', but rather: '/mnt/backup ' is REQUIRED
if ! grep -qs '/mnt/backup ' /proc/mounts; then
  echo 'ERROR! Backup location not mounted!'
  exit 1
fi

APP='/path/to/shrink-backup'
IMG_FILE='/mnt/backup/backup.img'

# Options
case "$1" in
  -a) "$APP" -Utlaz "$IMG_FILE";;
  -N) if [ -n "$2" ]; then
        "$APP" -tlz "$IMG_FILE" "$2"
      else
        "$APP" -atlz "$IMG_FILE"
      fi
      chown 1000:1000 "$IMG_FILE"
      ;;
  --loop) "$APP" -lz --loop "$IMG_FILE";;
  --fix) "$APP" -Utl --fix "$IMG_FILE" "$2";;
  --chroot) "$APP" -lz --chroot "$IMG_FILE";;
  *) "$APP" -Utlz "$IMG_FILE" "$1";;
esac

sleep 1

# Add compression here, change to compression method of your choice by editing the `tar` line
echo '## Compressing backup img'
tar -czvf /mnt/backup/backup.tar.gz /mnt/backup/backup.img
sleep 1

# Removing img file after compression is done
# If you update the img file, maybe you don't want this
# Or maybe you want to add a decompression of the tar.gz file at the start of this script
rm /mnt/backup/backup.img

exit 0
```

Make sure the script is executable with `chmod u+x /path/to/backup.sh`
And then run it with the option you prefer
```
    -N = New backup, (-atl options)
    -a = Update backup and autoresize if needed (-Utla options)
--loop = loop img file
--fix  = Update backup and fix "broken pipe" errors (-Utl --fix options)
```
Example: `sudo /path/to/backup.sh -N` will create a new backup at `/mnt/backup/backup.img` and then compress it into `/mnt/backup/backup.tar.gz` and finally remove the img file.
Running without any options, ie: `sudo /path/to/backup.sh` will result in an image update with zoom speed activated (* = `-Utlz` options)
You can obv edit/remove/add options to your liking.

This will probably be added in the wiki... at some point...
